core/domain/exp_domain.py

```diff
@@ -1611,6 +1611,57 @@ def create_default_state(
                 default_dest_state_name))
 
 
+class ExplorationVersionsDiff(object):
+    """Domain object for the difference between two versions of an Oppia
+    exploration.
+
+    Attributes:
+        states_added: list(str). The states added to the exploration from
+            prev_exp_version to current_exp_version.
+        states_deleted: list(str). The states deleted from the exploration from
+            prev_exp_version to current_exp_version.
+        states_renamed: dict. Dictionary mapping state names of
+            current_exp_version to the state names of prev_exp_version.
+    """
+
+    def __init__(self, change_list):
+        """Constructs an ExplorationVersionsDiff domain object.
+
+        Args:
+            change_list: list(dict). A list of all of the commit cmds from the
+                old_stats_model up to the next version.
+        """
+
+        states_added = []
+        states_deleted = []
+        states_renamed = {}
+
+        for change_dict in change_list:
+            if change_dict['cmd'] == CMD_ADD_STATE:
+                states_added.append(change_dict['state_name'])
+            elif change_dict['cmd'] == CMD_DELETE_STATE:
+                state_name = change_dict['state_name']
+                if state_name in states_added:
+                    states_added.remove(state_name)
+                else:
+                    states_deleted.append(state_name)
```

Comment:
Should we remove it from the "rename dict" in this case? 

What if a state gets renamed before it is deleted?

(Add backend tests for these cases too.)

Reply:
True. Fixed it differently, and tests cases added as well.